From 58fd4d2136c0639f06f82be04e4d7b4ba4ae3abd Mon Sep 17 00:00:00 2001
From: Alison Wang <alison.wang@nxp.com>
Date: Thu, 22 Nov 2018 09:37:06 +0800
Subject: [PATCH 375/706] drm: hdp: Add pixel engine reset for T28HPC

This patch adds pixel engine reset for T28HPC.

Signed-off-by: Alison Wang <alison.wang@nxp.com>
(cherry picked from commit 3de00148f4a83066e4dc2c78ec701851fe960684)
Signed-off-by: Zumeng Chen <zumeng.chen@windriver.com>
---
 drivers/gpu/drm/imx/hdp/imx-dp.c | 18 ++++++++++++++++++
 1 file changed, 18 insertions(+)

diff --git a/drivers/gpu/drm/imx/hdp/imx-dp.c b/drivers/gpu/drm/imx/hdp/imx-dp.c
index 339857794752..1855bb6bfe79 100644
--- a/drivers/gpu/drm/imx/hdp/imx-dp.c
+++ b/drivers/gpu/drm/imx/hdp/imx-dp.c
@@ -447,6 +447,21 @@ int dp_get_hpd_state(state_struct *state, u8 *hpd)
 	return ret;
 }
 
+void dp_phy_pix_engine_reset_t28hpc(state_struct *state)
+{
+	GENERAL_Read_Register_response regresp;
+
+	CDN_API_General_Read_Register_blocking(state, ADDR_SOURCE_CAR +
+					       (SOURCE_HDTX_CAR << 2),
+					       &regresp);
+	CDN_API_General_Write_Register_blocking(state, ADDR_SOURCE_CAR +
+						(SOURCE_HDTX_CAR << 2),
+						regresp.val & 0xFD);
+	CDN_API_General_Write_Register_blocking(state, ADDR_SOURCE_CAR +
+						(SOURCE_HDTX_CAR << 2),
+						regresp.val);
+}
+
 int dp_phy_init_t28hpc(state_struct *state,
 		       struct drm_display_mode *mode,
 		       int format,
@@ -460,6 +475,9 @@ int dp_phy_init_t28hpc(state_struct *state,
 	/* reset phy */
 	imx_hdp_call(hdp, phy_reset, 0);
 
+	dp_phy_pix_engine_reset_t28hpc(state);
+	DRM_INFO("pixel engine reset\n");
+
 	if (hdp->is_edp) {
 		max_link_rate = hdp->edp_link_rate;
 		num_lanes = hdp->edp_num_lanes;
-- 
2.17.1

